internal/controller/apiextensions/composite/composition_ptf.go

```diff
@@ -0,0 +1,982 @@
+	// Doing this before we patch and transform ensures we report the state we
+	// actually observed before we made any mutations.
+	o, err := FunctionIOObserved(state)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildFunctionIOObserved)
+	}
+
+	// Run P&T logic, updating the composition state accordingly.
+	if err := c.composition.PatchAndTransform(ctx, req, state); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errPatchAndTransform)
+	}
+
+	// Build the initial desired state to be passed to our Composition Function
+	// pipeline. It's expected that each function in the pipeline will mutate
+	// this state. It includes any desired state accumulated by the P&T logic.
+	d, err := FunctionIODesired(state)
+	if err != nil {
+		return CompositionResult{}, errors.Wrap(err, errBuildFunctionIODesired)
+	}
+
+	// Run Composition Functions, updating the composition state accordingly.
+	// Note that this will replace state.Composite with a new object that was
+	// unmarshalled from the function pipeline's desired state.
+	if err := c.composition.RunFunctionPipeline(ctx, req, state, o, d); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errRunFunctionPipeline)
+	}
+
+	// Garbage collect any resources that aren't part of our final desired
+	// state. We must do this before we update the XR's resource references to
+	// ensure that we don't forget and leak them if a delete fails.
+	if err := c.composite.DeleteComposedResources(ctx, state); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errDeleteUndesiredCDs)
+	}
+
+	// Record references to all desired composed resources.
+	UpdateResourceRefs(state)
+
+	// The supplied options ensure we merge rather than replace arrays and
+	// objects for which a merge configuration has been specified.
+	//
+	// Note that at this point state.Composite should be a new object - not the
+	// xr that was passed to this Compose method. If this call to Apply changes
+	// the XR in the API server (i.e. if it's not a no-op) the xr object that
+	// was passed to this method will have a stale meta.resourceVersion. This
+	// Subsequent attempts to update that object will therefore fail. This
+	// should be okay; the caller should keep trying until this is a no-op.
+	ao := mergeOptions(filterPatches(allPatches(state.ComposedResources), patchTypesToXR()...))
+	if err := c.client.Apply(ctx, state.Composite, ao...); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errApplyXR)
+	}
+
+	// We apply all of our composed resources before we observe them and update
+	// in the loop below. This ensures that issues observing and processing one
+	// composed resource won't block the application of another.
+	for _, cd := range state.ComposedResources {
+		// Don't try to apply this resource if we didn't render it successfully.
+		// Note that this doesn't mean this resource won't exist; it might have
+		// been created previously.
+		if !cd.Rendered {
```

Comment:
I pushed https://github.com/crossplane/crossplane/pull/3465/commits/5c3e6772fc39709b8c63ceb5df213668d0c7cb7e to fix this.

>  I think we can consider the function pipeline itself to be "rendering" composed resources, and potentially have `ParseDesiredResource` set `Rendered == true`.

This approach didn't work out for a few reasons. At the moment we pass all resources from the P&T process as the initial desired state of the function pipeline, even if those resource failed to render. So the function pipeline might be supplied with broken or partial desired state.

If `ParseDesiredResources` just set `Rendered == true` for every desired resource that was produced by the pipeline it would 'reset' these broken resources that were passed into it from P&T Composition back to `Rendered == true` and we'd try apply them.

I considered that maybe we shouldn't pass these partially rendered resources into the function pipeline in the first place. After all, functions won't have any way to distinguish an existing desired resource that is correct, fully formed, and will apply successfully from one that isn't. To be fair, this is something functions are going to have to tolerate regardless - for example they'd need to tolerate a previous function in the pipeline producing a desired composed resource that might not be able to be applied successfully.

The issue with not passing P&T composed resources that didn't render successfully to the function pipeline is that we currently consider any resource that didn't 'survive' the function pipeline to be eligible for garbage collection. Because we fetch any existing composed resources first this means it's possible that we'd erroneously garbage collect an existing composed resource that experienced a temporary P&T render error.

I don't particularly love the asymmetry between P&T and the function pipeline here:

* If P&T fails to render something, we remember and don't apply it.
* If a function fails to render something it's expected to omit it from its desired state.

I considered just removing the check in the apply loop entirely, and just trying to apply everything even if we know the render failed. In the likely event that the apply also failed we could just emit a warning event and try again on the next reconcile. This is the approach I like most, but I think where it falls apart is the edge case in which:

1. We partially render a composed resource, then hit an error.
2. The resource is fully formed enough to apply successfully.
3. We later successfully render it, but can't update it to its "successful" state (e.g. due to an immutable field).